builtin/game/falling.lua

```diff
@@ -53,81 +52,92 @@ core.register_entity(":__builtin:falling_node", {
+		local below_pos = {x = pos.x, y = pos.y - 0.7, z = pos.z}
+		-- Avoid bugs caused by an unloaded node below
+		local below_node = core.get_node(below_pos)
 		-- Delete on contact with ignore at world edges
-		if bcn and bcn.name == "ignore" then
+		if below_node.name == "ignore" then
 			self.object:remove()
 			return
 		end
-		local bcd = bcn and core.registered_nodes[bcn.name]
-		if bcn and
-				(not bcd or bcd.walkable or
-				(core.get_item_group(self.node.name, "float") ~= 0 and
-				bcd.liquidtype ~= "none")) then
-			if bcd and bcd.leveled and
-					bcn.name == self.node.name then
-				local addlevel = self.node.level
-				if not addlevel or addlevel <= 0 then
-					addlevel = bcd.leveled
-				end
-				if core.add_node_level(bcp, addlevel) == 0 then
-					self.object:remove()
-					return
-				end
-			elseif bcd and bcd.buildable_to and
-					(core.get_item_group(self.node.name, "float") == 0 or
-					bcd.liquidtype == "none") then
-				core.remove_node(bcp)
+
+		local below_nodef = core.registered_nodes[below_node.name]
+		-- Is it a level node we can add to?
+		if below_nodef and below_nodef.leveled and
+				below_node.name == self.node.name then
+			local addlevel = self.node.level
+			if not addlevel or addlevel <= 0 then
+				addlevel = below_nodef.leveled
+			end
+			if core.add_node_level(below_pos, addlevel) == 0 then
+				self.object:remove()
 				return
 			end
-			local np = {x = bcp.x, y = bcp.y + 1, z = bcp.z}
-			-- Check what's here
-			local n2 = core.get_node(np)
-			local nd = core.registered_nodes[n2.name]
-			-- If it's not air or liquid, remove node and replace it with
-			-- it's drops
-			if n2.name ~= "air" and (not nd or nd.liquidtype == "none") then
-				core.remove_node(np)
-				if nd and nd.buildable_to == false then
-					-- Add dropped items
-					local drops = core.get_node_drops(n2, "")
+		end
+
+		-- Stop node if it falls on walkable surface, or floats on water
+		if (below_nodef and below_nodef.walkable == true)
+				or (below_nodef
+				and core.get_item_group(self.node.name, "float") ~= 0
+				and below_nodef.liquidtype ~= "none") then
+			self.object:set_velocity({x = 0, y = 0, z = 0})
+		end
+
+		-- Has the fallen node stopped moving ?
+		local vel = self.object:get_velocity()
+		if vector.equals(vel, {x = 0, y = 0, z = 0}) then
+			local npos = self.object:get_pos()
+			-- Get node we've landed inside
+			local cnode = minetest.get_node(npos).name
+			local cdef = core.registered_nodes[cnode]
+			-- If airlike or buildable_to or an attached_node then place node,
+			-- otherwise drop falling node as an item instead.
+			if (cdef and cdef.air_equivalent == true)
+					or (cdef and cdef.buildable_to == true)
+					or (cdef and cdef.liquidtype ~= "none")
+					or core.get_item_group(cnode, "attached_node") ~= 0 then
+				-- Are we an attached node ? (grass, flowers, torch)
+				if core.get_item_group(cnode, "attached_node") ~= 0 then
+					-- Add drops from attached node
```

Comment:
Previously drops were added when the removed node was not air, not liquid and not buildable_to.
Now drops are added only for attached nodes, a change of behaviour.